File: sdk/go/sdk/buildkite/group_step.go
package buildkite

import "github.com/buildkite/buildkite-sdk/sdk/go/sdk/schema"

type GroupStepStep interface {
	ToPipelineStep() *PipelineStep
}

type GroupStep struct {
	DependsOn              DependsOn       `json:"depends_on,omitempty"`
	Group                  *string         `json:"group,omitempty"`
	If                     *string         `json:"if,omitempty"`
	Key                    *string         `json:"key,omitempty"`
	ID                     *string         `json:"id,omitempty"`
	Identifier             *string         `json:"identifier,omitempty"`
	Label                  *string         `json:"label,omitempty"`
	Name                   *string         `json:"name,omitempty"`
	AllowDependencyFailure *bool           `json:"allow_dependency_failure,omitempty"`
	Skip                   *bool           `json:"skip,omitempty"`
	Notify                 []StepNotify    `json:"notify,omitempty"`
	Steps                  []GroupStepStep `json:"steps,omitempty"`
}

func (step GroupStep) ToPipelineStep() *PipelineStep {
	groupName := Value("~")
	if step.Group != nil {
		groupName = step.Group
	}

	notify := make([]schema.BlockStepNotify, len(step.Notify))
	for i, val := range step.Notify {
		notify[i] = *val.toSchema()
	}

	steps := make([]PipelineStep, len(step.Steps))
	for i, step := range step.Steps {
		steps[i] = *step.ToPipelineStep()
	}

	groupStep := &PipelineStep{
		Group:      groupName,
		If:         step.If,
		ID:         step.ID,
		Identifier: step.Identifier,
		Key:        step.Key,
		Label:      step.Label,
		Name:       step.Name,
		Notify:     notify,
		Steps:      steps,
	}

	if step.AllowDependencyFailure != nil {
		groupStep.AllowDependencyFailure = &schema.AllowDependencyFailureUnion{
			Bool: step.AllowDependencyFailure,
		}
	}

	if step.DependsOn != nil {
		groupStep.DependsOn = step.DependsOn.toSchema()
	}

	if step.Skip != nil {
		groupStep.Skip = &schema.Skip{
			Bool: step.Skip,
		}
	}

	return groupStep
}

File: sdk/csharp/src/Buildkite.Sdk/Schema/GroupStep.cs
using System.Text.Json.Serialization;
using YamlDotNet.Serialization;

namespace Buildkite.Sdk.Schema;

/// <summary>
/// A group step organizes steps into a collapsible group.
/// </summary>
public class GroupStep : IStep
{
    /// <summary>The label for the group.</summary>
    public string? Group { get; set; }

    /// <summary>Alias for Group.</summary>
    [JsonIgnore]
    [YamlIgnore]
    public string? Label { get => Group; set => Group = value; }

    /// <summary>Alias for Group.</summary>
    [JsonIgnore]
    [YamlIgnore]
    public string? Name { get => Group; set => Group = value; }

    /// <summary>A unique identifier for this step.</summary>
    public string? Key { get; set; }

    /// <summary>Alias for Key.</summary>
    [JsonIgnore]
    [YamlIgnore]
    [Obsolete("Use Key instead.")]
    public string? Id { get => Key; set => Key = value; }

    /// <summary>Alias for Key.</summary>
    [JsonIgnore]
    [YamlIgnore]
    public string? Identifier { get => Key; set => Key = value; }

    /// <summary>The steps within this group.</summary>
    public List<IGroupStep> Steps { get; set; } = new();

    /// <summary>A boolean expression to conditionally run this step.</summary>
    public string? If { get; set; }

    /// <summary>Glob pattern to run step only if matching files changed.</summary>
    public object? IfChanged { get; set; }

    /// <summary>Step keys this step depends on.</summary>
    public object? DependsOn { get; set; }

    /// <summary>Whether to proceed if a dependency fails.</summary>
    public bool? AllowDependencyFailure { get; set; }

    /// <summary>Whether to skip this step.</summary>
    public object? Skip { get; set; }

    /// <summary>Notifications for this group.</summary>
    public List<INotification>? Notify { get; set; }

    /// <summary>Add a step to this group.</summary>
    public GroupStep AddStep(IGroupStep step)
    {
        Steps.Add(step);
        return this;
    }
}
